libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/SharedApp.kt

```diff
@@ -0,0 +1,37 @@
+package com.mapbox.navigation.ui.app.internal
+
+import android.content.Context
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
+import com.mapbox.navigation.ui.utils.internal.configuration.NavigationConfigOwner
+import com.mapbox.navigation.ui.utils.internal.datastore.NavigationDataStoreOwner
+import com.mapbox.navigation.ui.voice.internal.MapboxAudioGuidance
+import com.mapbox.navigation.ui.voice.internal.impl.MapboxAudioGuidanceImpl
+import com.mapbox.navigation.ui.voice.internal.impl.MapboxAudioGuidanceServicesImpl
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+object SharedApp {
```

Comment:
What do you think about making this something like `object SharedServices : MapboxNavigationObserver`

And then it does not require the context to be passed. That is provided by `MapboxNavigationApp.setup`. Both AA and DropIn can include this in their creation. It makes it so, developers 

``` kotlin
class NavigationView
...
  init { 
    MapboxNavigationApp.registerObserver(MapboxSharedAppServices)
```
```kotlin
class MainCarSession : Session() {
  init {
    MapboxNavigationApp.registerObserver(MapboxSharedAppServices)
```

Reply:
Was also considering this one. But this is more fragile when a developer calls `MapboxNavigationApp.setup` to create their own `NavigationOptions`

``` kotlin
        if (!MapboxNavigationApp.isSetup()) {
            MapboxNavigationApp.setup(
                NavigationOptions.Builder(context)
                    .accessToken(accessToken)
                    .build()
            ).registerObserver(MapboxSharedAppServices)
        }
```

Reply:
I thought about it but unfortunately, that didn't guarantee that the `SharedApp.setup()` code will be executed before any other components from DropIn and AndroidAuto are.

Reply:
Where's the issue exactly? This class is very similar to `MapboxNavigationApp`. The goal of [this change](https://github.com/mapbox/mapbox-navigation-android/pull/5829) is to provide low level support for what this `SharedApp` is doing. Because it would be nice to be able to make `MapboxCarApp` and `SharedApp` unnecessary. 

What about something like this in the libnavui-voice module
``` kotlin
fun MapboxNavigationApp.registerAudioGuidance(context: Context) {
    if (getObservers(MapboxAudioGuidance::class).isEmpty()) {
        val mapboxAudioGuidance = MapboxAudioGuidanceImpl(
            MapboxAudioGuidanceServicesImpl(),
            NavigationConfigOwner(context)
        ).also {
            it.dataStoreOwner = NavigationDataStoreOwner(context, "mapbox_navigation_preferences")
        }
        registerObserver(mapboxAudioGuidance)
    }
}

@ExperimentalPreviewMapboxNavigationAPI
fun MapboxNavigationApp.unregisterAudioGuidance() {
    getObservers(MapboxAudioGuidance::class).forEach {
        unregisterObserver(it)
    }
}
```

There still is going to be a problem when we try to share `NavigationConfigOwner` between multiple observers. But that is an orthogonal issue.

Reply:
Instead of an extra set up call, what if it was a lazy loading factory. End of the day, this is a lazy load. What if it was only created where it is needed.

``` kotlin
AudioGuidanceStateController
..
- val audioGuidance = MapboxNavigationApp.getObserver(MapboxAudioGuidance::class)
+ val audioGuidance = MapboxNavigationApp.getAudioGuidance()
```


``` kotlin
@ExperimentalPreviewMapboxNavigationAPI
fun MapboxNavigationApp.getAudioGuidance(context: Context): MapboxAudioGuidance {
    val factory: (Context) -> MapboxAudioGuidance = { factoryContext ->
        MapboxAudioGuidanceImpl(
            MapboxAudioGuidanceServicesImpl(),
            NavigationConfigOwner(factoryContext)
        ).also {
            it.dataStoreOwner = NavigationDataStoreOwner(factoryContext, "mapbox_navigation_preferences")
            registerObserver(it)
        }
    } 
    return getObservers(MapboxAudioGuidance::class).firstOrNull() ?: factory(context)
}
```

The context required by NavigationConfigOwner` would need to be resolved to remove the context from the getAudioGuidance function

Reply:
Sorry, for iterating on ideas in this comment thread 😅